dotcom-rendering/src/web/components/Liveness.importable.tsx

```diff
@@ -0,0 +1,186 @@
+}
+
+/**
+ * revealNewBlocks - style any blocks that have been inserted but are hidden such that
+ * they are revealed
+ */
+function revealNewBlocks() {
+	console.log('revealNewBlocks');
+}
+
+/**
+ * This abstracts the code to construct the url we use for polling. Of note is the fact
+ * we have put `latestBlockId` in react state; by doing this we are able to change the url
+ * that SWR uses for the request whilst still benefitting from the polling logic
+ *
+ * @returns string
+ */
+function getKey(
+	pageId: string,
+	ajaxUrl: string,
+	filterKeyEvents: boolean,
+	latestBlockId: string,
+): string {
+	// Construct the url to poll
+	const url = new URL(`${pageId}.json`, ajaxUrl);
+	url.searchParams.set('lastUpdate', latestBlockId);
+	url.searchParams.set('isLivePage', 'true');
+	url.searchParams.set('filterKeyEvents', filterKeyEvents ? 'true' : 'false');
+	return url.href;
+}
+
+const topOfBlog: Element | null | false =
+	!isServer && window.document.querySelector('[data-gu-marker=top-of-blog]');
+
+/**
+ * This allows us to make decisions in javascript based on if the reader
+ * has the top of the blog in view or not
+ *
+ * @returns boolean
+ */
+function topOfBlogVisible(): boolean {
+	return !!topOfBlog && !!topOfBlog.classList.contains('in-viewport');
+}
+
+export const Liveness = ({ pageId, webTitle, ajaxUrl }: Props) => {
+	const [showToast, setShowToast] = useState(false);
+	const [noOfNewPosts, setNoOfNewPosts] = useState(0);
+	const [latestBlockId, setLatestBlockId] = useState(
+		// By default we use the first (latest) block id on the page
+		document.querySelector('#maincontent :first-child')?.id || '',
+	);
+
+	// Read current url to get filterKeyEvents
+	const params = new URLSearchParams(window.location.search);
+	const filterKeyEvents = !!params.get('filterKeyEvents');
+
+	// useApi returns { data, loading, error } but we're not using them here
+	useApi<{
+		numNewBlocks: number;
+		html: string;
+	}>(getKey(pageId, ajaxUrl, filterKeyEvents, latestBlockId), {
+		refreshInterval: 15000,
+		refreshWhenHidden: true,
+		// onSuccess runs (once) after every successful api call. This is useful because it
+		// allows us to avoid the problems of imperative code being executed multiple times
+		// inside react's declarative structure (things get re-rendered when any state changes)
+		onSuccess: (data: {
+			numNewBlocks: number;
+			html: string;
+			mostRecentBlockId: string;
+		}) => {
+			if (data && data.numNewBlocks && data.numNewBlocks > 0) {
+				// Always insert the new blocks in the dom (but hidden)
+				insertNewBlocks(data.html);
+				hydrateBlocks();
+
+				if (topOfBlogVisible()) {
+					revealNewBlocks();
+					setNoOfNewPosts(0);
+				} else {
+					setShowToast(true);
+					// Increment the count of new posts
+					setNoOfNewPosts(noOfNewPosts + data.numNewBlocks);
+				}
+
+				// Update the block id we use for polling
+				if (data.mostRecentBlockId)
+					setLatestBlockId(data.mostRecentBlockId);
```

Comment:
Do you have any thoughts on the use of the ESLint "curly" rule? https://eslint.org/docs/rules/curly

```suggestion
				if (data.mostRecentBlockId) {
					setLatestBlockId(data.mostRecentBlockId);
				}
```

Reply:
I think in this example using the braces like this is much better 👍 .

The lint rule though would affect quite a lot of code in DCR where adding braces would - in my opinion - interrupt the flow. Code like

```
if (someError) return <Error />
if (loading) return <Loading />

return (
  ...
)
```

would get messy. I think it's nice that, as it stands, we have the choice to add them if we think they work but they are not enforced unless you have multiple lines.

Now that we have Prettier, these days I'm generally happy to sheep-like follow whatever lint rules a project has. I just write code, save, and let Prettier do it's thing. I really like how this tool removes the need for all those styling discussions we used to have before it existed!